contrib/mpi-proxy-split/mpi-wrappers/mpi_collective_p2p.c

```diff
@@ -0,0 +1,630 @@
+    recvcounts = sendcounts;
+    recvtype = sendtype;
+    // rdisps will not be used.
+  }
+  MPI_Aint lower_bound;
+  MPI_Aint sendextent, recvextent;
+  MPI_Type_get_extent(sendtype, &lower_bound, &sendextent);
+  if (inplace) {
+    recvextent = sendextent;
+  } else {
+    MPI_Type_get_extent(recvtype, &lower_bound, &recvextent);
+  }
+  // FIXME:  Add assert: Sum of sendcounts[]*extent(sendtype) == SAME_FOR_RECV
+  // assert(sendextent*sendcount == recvextent*recvcount);
+  for (i = 0; i < size; i++) {
+    if (i == rank) {
+      if (!inplace) {
+        memcpy(recvbuf + rdispls[i]*recvextent /* i*recvextent*recvcount */,
+               sendbuf + sdispls[i]*sendextent /* i*sendextent*sendcount */,
+               sendextent*sendcounts[i]);
+      } // NOTE: if inplace, MPI guarantees that rank's data is already correct
+    } else { // else: i != rank
+      MPI_Send(sendbuf + sdispls[i]*sendextent /* sdispls[i] == i*sendcount */,
+               sendcounts[i], sendtype,
+               i, 0, comm);
+    }
+  }
+  for (i = 0; i < size; i++) {
+    // NOTE: if i == rank, we handled it during MPI_Send
+    if (i != rank) {
+      const int *displs = (inplace ? sdispls : rdispls);
+      MPI_Recv(recvbuf + displs[i]*recvextent /* rdispls[i] == i*recvcount */,
+               recvcounts[i], recvtype,
+               i, 0, comm, MPI_STATUS_IGNORE);
+    }
+  }
+  return MPI_SUCCESS;
+}
+
+int MPI_Alltoallw(const void* sendbuf, const int sendcounts[],
+                  const int sdispls[], const MPI_Datatype sendtypes[],
+                  void* recvbuf, const int recvcounts[], const int rdispls[],
+                  const MPI_Datatype recvtypes[], MPI_Comm comm) {
+  fprintf(stderr, "%s not implemented\n", __FUNCTION__);
+  ABORT();
+  return -1;
+}
+
+// MPI standard 3.1:  Section 5.9
+/* NOTE:  MPI-3.1 standard (Section 5.9.1):
+ *   It is strongly recommended that MPI_REDUCE be implemented so that the
+ *   same result be obtained whenever the function is applied on the same
+ *   arguments, appearing in the same order.
+ */
+// Predefined Reduction Operatoins (Section 5.9.2)
+/*
+ * MPI_MAX
+ * MPI_MIN
+ * MPI_SUM
+ * MPI_PROD
+ * MPI_LAND
+ * MPI_BAND
+ * MPI_LOR
+ * MPI_BOR
+ * MPI_LXOR
+ * MPI_BXOR
+ * MPI_MAXLOC
+ * MPI_MINLOC
+ */
+int MPI_Reduce(const void* sendbuf, void* recvbuf, int count,
+               MPI_Datatype datatype, MPI_Op op, int root, MPI_Comm comm) {
+  PROLOG_rank_size;
+  MPI_Aint lower_bound;
+  MPI_Aint extent;
+  MPI_Type_get_extent(datatype, &lower_bound, &extent);
+  int inplace = (sendbuf == MPI_IN_PLACE);
+  if (inplace && rank == root) {
+    sendbuf = recvbuf;
+  }
+
+  if (rank == root) {
+    // Gather data into tmp_sendbuf at root
+    int i;
+    char *tmp_sendbuf = (char *)malloc(count * extent);
+    MPI_Gather(sendbuf, count, datatype, tmp_sendbuf, count, datatype,
+               root, comm);
+    // Initialize tmp_recvbuf from sendbuf at rank 0
+    char *tmp_recvbuf = malloc(count * extent); 
```

Comment:
Unfortunately, g++ doesn't allow implicit conversion.
gcc allows it, but this file was included in cpp so will be compiled by g++